### PR TITLE
fpm deployment enhancement

### PR DESCRIPTION
**Pull request by jalvesz**
#791 #790 #787 

Enable building stdlib with fpm directly from the root folder
> python config/fypp_deployment.py
> fpm build

Deploy the `stdlib-fpm` from the preprocessing created in the root folder. This deployment could be rendered unnecessary by building directly stdlib from the root.

### CODE REVIEW

**Comment by jalvesz** [comment]
@perazz maybe you know if there is a way to avoid launching the tests `test_always_fail` and `test_always_skip` when doing `fpm test` ? The other solution I have in mind is to explicitly add the tests to the manifest and deactivate the auto-test option?

**Comment by perazz** [comment]
Yes I fear that would be the only option, though it's clearly sub-optimal (for each new test, it will have to be added to `fpm.toml`). I think the question is why those tests are there in the first place for? In other words, is there a valid reason those dummy tests should be kept into `stdlib`, or they could be safely removed? At the end of the day they are testing `test-drive` rather than stdlib.

**Comment by jalvesz** [comment]
>I think the question is why those tests are there in the first place for? In other words, is there a valid reason those dummy tests should be kept into stdlib, or they could be safely removed?

I was wondering the same thing!

**Comment by jvdp1** [comment]
> > I think the question is why those tests are there in the first place for? In other words, is there a valid reason those dummy tests should be kept into stdlib, or they could be safely removed?
> 
> I was wondering the same thing!

These tests were introduced for testing the procedure `check` in `stdlib_error`. Later we started to use `test-drive`.
If `check` is not used anymore, I guess these tests could be removed.

**Comment by jalvesz** [comment]
Thanks @jvdp1, removing those tests enables indeed testing from the root build using fpm!

Before going further I wanted to ask about the `stdlib-fpm` branch. Given that this mechanism enables preprocessing and building from root, should this branch be maintained nonetheless? or shall we envision deprecating it progressively? Just to know "how far" should we push this build mechanism.

I'll propose a modification to the README.md before moving from draft to ready for review, but wanted to have some feedback on this point.

File: README.md [hunk 1]
```diff
@@ -179,19 +179,38 @@ earlier builds do not affect the new build.
 Fortran Package Manager (fpm) is a package manager and build system for Fortran.   
 You can build `stdlib` using provided `fpm.toml`:
 
+**Option 1**: From root folder
+
+As `fpm` does not currently support `fypp` natively, `stdlib` now proposes a python script to preprocess and build. This script enables modification of the different `fypp` macros available in `stdlib`. The preprocessed files will be dumped at `<current_folder>/temp/*.f90` or `*.F90`. You can use the following command line:
```
Comment (jvdp1) [inline README.md:184]:
```suggestion
As `fpm` does not currently support `fypp` natively, `stdlib` now proposes a python script to preprocess and build it.
This script enables modification of the different `fypp` macros available in `stdlib`. The preprocessed files will be dumped at `<current_folder>/temp/*.f90` or `*.F90`.
You can use the following command line:
```

File: README.md [hunk 2]
```diff
@@ -179,19 +179,38 @@ earlier builds do not affect the new build.
...
 ```
 
-**Alternative**: as `fpm` does not currently support `fypp` natively, building `stdlib` with `fpm` can be done in two steps: a) launch the preprocessor through the `fpm-deployment.sh` script, which creates a subfolder `stdlib-fpm` and b) build the project using the processed files within the latter subfolder. This process can be done with the following commands:
+or the short-cut
 
 ```sh
-source ./ci/fpm-deployment.sh
-cd stdlib-fpm/
-fpm build --profile release
+python config/fypp_deployment.py --build 1
```
Comment (jvdp1) [inline README.md:194]: Why is the `1` needed?
could it be:
```suggestion
python config/fypp_deployment.py --build
```

File: config/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+fypp
```
Comment (jvdp1) [inline config/requirements.txt:1]: Why is this file needed? What is its aim?

**Review by jvdp1 — COMMENTED** [review]
LGTM. Thank you @jalvesz . I just have a few last comments.
A nice improvement is the support of the cpp directives in the branch `stdlib-fpm`.

File: README.md (hunk 2 above)
Comment (jalvesz) [reply README.md:194]: Since I defined the parameter as a boolean, it expects a value (an int that can be 0 or 1) after the key word, otherwise it throws an error, I can check if there is a way to facilitate that.

File: config/requirements.txt (hunk 3 above)
Comment (jalvesz) [reply config/requirements.txt:1]: Since we need here python for the preprocessing, I added the requirements file for installation of dependencies. Maybe I should add that to the README

File: README.md (hunk 2 above)
Comment (jalvesz) [reply README.md]: found it https://stackoverflow.com/questions/8259001/python-argparse-command-line-flags-without-arguments and solved with this commit https://github.com/fortran-lang/stdlib/pull/802/commits/95ee064fa3d61d5b6d4107bface789af6d6ae462

File: config/fypp_deployment.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
+import os
+import fypp
+import argparse
+from joblib import Parallel, delayed
```
Comment (jvdp1) [inline config/fypp_deployment.py:4]: I am just trying the script now on a different machine. It could be good to mention somewhere the these should be installed first (if it is not already the case, of course).

Comment (jalvesz) [reply config/fypp_deployment.py:4]: Oh yes, that was the main purpose of the requirements.txt file, to simply do `pip install -r requirements.txt`. I'll add those and the comment in the README.

Comment (jalvesz) [reply config/fypp_deployment.py:4]: Done on commit https://github.com/fortran-lang/stdlib/pull/802/commits/a5978340d587361e6ffded32f8f5715740fbc20b

**Review by jvdp1 — APPROVED** [review]
I played a bit with the script, and I think it adds nice features (especially that `cpp` directives are still kept in `stdlib-fpm`).

Thank you @jalvesz!

**Review by perazz — APPROVED** [review]
Looks great to me in its current form. 
Down the road, I think we may want this script to be able to pass whatever defined preprocessor macros to the `fpm.toml`: 

```toml
[preprocess]
[preprocess.cpp]
macros = ["added","by","deployment","script"]
```
So, let's keep that in mind for now.

**Comment by jalvesz** [comment]
> Down the road, I think we may want this script to be able to pass whatever defined preprocessor macros to the fpm.toml:

Totally! While in the current PR I intentionally striped out that function to make things concise, in the script I posted in #791 that function is still available and can be easily added.

**Comment by jvdp1** [comment]
@jalvesz @perazz It seems that this PR is ready. I will merge it in a couple of days.
